File: typescript4/src/comparison.ts
import { Factory } from './construction';
import * as serialization from './serialization';
import * as traversal from './traversal';

type Result = -1 | 0 | 1 | undefined;

class Comparator<ExpectedType extends serialization.Serializable> implements traversal.Visitor<ExpectedType> {
    a: any;
    b: any;
    result: Result;

    constructor(a: ExpectedType, b: ExpectedType) {
        this.a = a;
        this.b = b;
        this.result = undefined;
    }

    begin(obj: ExpectedType, parentPropName?: string): Result {
        this.result = 0;
        return this.result;
    }

    end(obj: ExpectedType): Result {
        return this.result;
    }

    owner(target: ExpectedType, ownerPropName: string) {        
    }

    verbatim<DataType>(
        target: serialization.Serializable,
        getValue: (target: serialization.Serializable) => any,
        setValue: (target: serialization.Serializable, value: any) => void
    ): Result {
        if(this.result !== 0) {
            return this.result;
        }
        
        const aProp = getValue(this.a);
        const bProp = getValue(this.b);
        if(aProp === undefined && bProp !== undefined) {
            this.result = -1;
        } else if(aProp !== undefined && bProp === undefined) {
            this.result = 1;
        } else if(aProp === undefined && bProp === undefined) {
            // pass
        } else if(aProp < bProp) {
            this.result = -1;
        } else if(aProp > bProp) {
            this.result = 1;
        }
        return this.result;
    }

    primitive<PropType>(target: any, propName: string, fromString?: (initializer:string) => PropType): Result {
        if(this.result !== 0) {
            return this.result;
        }

        const aHasProp = this.a.hasOwnProperty(propName);
        const bHasProp = this.b.hasOwnProperty(propName);
        if(!aHasProp && !bHasProp) {
            return this.result;
        } else if(!aHasProp && bHasProp) {
            this.result = -1;
        } else if(aHasProp && !bHasProp) {
            this.result = 1;
        }

        const aProp = this.a[propName];
        const bProp = this.b[propName];
        if(aProp === undefined && bProp !== undefined) {
            this.result = -1;
        } else if(aProp !== undefined && bProp === undefined) {
            this.result = 1;
        } else if(aProp === undefined && bProp === undefined) {
            // pass
        } else if(aProp < bProp) {
            this.result = -1;
        } else if(aProp > bProp) {
            this.result = 1;
        }
        return this.result;
    }

    reference<ElementType extends serialization.Serializable>(
        elementFactory: Factory<ElementType>,
        propName: string
    ): Result {
        if(this.result !== 0) {
            return this.result;
        }

        const aHasProp = this.a.hasOwnProperty(propName);
        const bHasProp = this.b.hasOwnProperty(propName);
        if(!aHasProp && !bHasProp) {
            return this.result;
        } else if(!aHasProp && bHasProp) {
            this.result = -1;
        } else if(aHasProp && !bHasProp) {
            this.result = 1;
        }

        const aProp = this.a[propName];
        const bProp = this.b[propName];
        if(aProp._ref < bProp._ref) {
            this.result = -1;
        } else if(aProp._ref > bProp._ref) {
            this.result = 1;
        } else if(aProp._def < bProp._def) {
            this.result = -1;
        } else if(aProp._def > bProp._def) {
            this.result = 1;
        }
        return this.result;
    }

    compare_elements<ElementType extends serialization.Serializable>(
        a_prop: serialization.Serializable,
        b_prop: serialization.Serializable
    ) {
        if(a_prop.getGlobalId() === null && b_prop.getGlobalId() === null) {
            const sub = new Comparator(a_prop, b_prop);
            a_prop.marshal(sub);
            if(sub.result !== 0) {
                this.result = sub.result;
            }
        } else if(a_prop.getGlobalId() === null && b_prop.getGlobalId() !== null) {
            this.result = -1;
        } else if(a_prop.getGlobalId() !== null && b_prop.getGlobalId() === null) {
            this.result = 1;
        } else if((a_prop.getGlobalId() || 0) < (b_prop.getGlobalId() || 0)) {
            this.result = -1;
        } else if((a_prop.getGlobalId() || 0) >(b_prop.getGlobalId() || 0)) {
            this.result = -1;
        }
    }

    scalar<ElementType extends serialization.Serializable>(
        elementFactory: Factory<ElementType>,
        propName: string
    ): Result {
        if(this.result !== 0) {
            return this.result;
        }

        const aHasProp = this.a.hasOwnProperty(propName);
        const bHasProp = this.b.hasOwnProperty(propName);
        if(!aHasProp && !bHasProp) {
            return this.result;
        } else if(!aHasProp && bHasProp) {
            this.result = -1;
        } else if(aHasProp && !bHasProp) {
            this.result = 1;
        }

        const aProp = this.a[propName];
        const bProp = this.b[propName];
        if(aProp === undefined && bProp === undefined) {
            return this.result;
        } else if(aProp === undefined && bProp !== undefined) {
            this.result = -1;
        } else if(aProp !== undefined && bProp === undefined) {
            this.result = 1;
        } else {
            this.compare_elements(aProp, bProp);
        }
        if(this.result === undefined) {
            this.result = 0;
        }
        return this.result;
    }

    array<ElementType extends serialization.Serializable>(
        elementFactory: Factory<ElementType>,
        propName: string
    ): Result {
        const aHasProp = this.a.hasOwnProperty(propName);
        const bHasProp = this.b.hasOwnProperty(propName);
        if(!aHasProp && !bHasProp) {
            return this.result;
        } else if(!aHasProp && bHasProp) {
            this.result = -1;
        } else if(aHasProp && !bHasProp) {
            this.result = 1;
        }
        if(this.a.length < this.b.length) {
            this.result = -1;
        } else if(this.a.length > this.b.length) {
            this.result = 1;
        }

        this.a[propName].forEach((aProp: serialization.Serializable, index: number) => {
            const bProp = this.b[propName][index];
            if(this.result === undefined) {
                this.compare_elements(aProp, bProp);
            }
        });
        if(this.result === undefined) {
            this.result = 0;
        }
        return this.result;
    }

    map<ElementType extends serialization.Serializable>(
        elementFactory: Factory<ElementType>,
        propName: string
    ): Result {
        const aHasProp = this.a.hasOwnProperty(propName);
        const bHasProp = this.b.hasOwnProperty(propName);
        if(!aHasProp && !bHasProp) {
            return this.result;
        } else if(!aHasProp && bHasProp) {
            this.result = -1;
        } else if(aHasProp && !bHasProp) {
            this.result = 1;
        }
        const aKeys = Object.getOwnPropertyNames(this.a[propName]);
        const bKeys = Object.getOwnPropertyNames(this.b[propName]);
        if(aKeys < bKeys) {
            this.result = -1;
        } else if(aKeys > bKeys) {
            this.result = 1;
        }
        aKeys.forEach((key: string) => {
            if(this.result === undefined) {
                this.compare_elements(this.a[propName][key], this.b[propName][key]);
            }
        });
        if(this.result === undefined) {
            this.result = 0;
        }
        return this.result;
    }
}

export function cmp<DataType>(a: DataType|undefined, b:DataType|undefined, comparator?: (a: DataType, b: DataType) => Result): Result {
    if(a === undefined || b === undefined) {
        if(a === undefined && b !== undefined) {
            return -1;
        } else if(a !== undefined&& b === undefined) {
            return 1;
        }
        return 0;
    }

    if(a === null || b === null) {
        if(a === null && b !== null) {
            return -1;
        } else if(a !== null&& b === null) {
            return 1;
        }
        return 0;
    }

    if(a instanceof serialization.Serializable && b instanceof serialization.Serializable) {
        const comparator = new Comparator(a,b);
        a.marshal(comparator);
        return comparator.result;
    }

    if(comparator) return comparator(a,b);

    const use_native_comparator = ['string', 'number', 'boolean', 'object'];
    if(use_native_comparator.includes(typeof(a)) && use_native_comparator.includes(typeof(b))) {
        if(a < b) {
            return -1;
        } else if(a > b) {
            return 1;
        }
        return 0;
    }

    return undefined;
}

File: python3/elevated_objects/comparison.py
#!/usr/bin/env python3

from __future__ import annotations
import typing
import enum
import sys

from code_instruments.task import function, method, Tally
from . import construction
from . import serializable
from . import traversal

PropType = typing.TypeVar('PropType', bound=serializable.Serializable)

#logger=lambda x: sys.stdout.write(x+"\n")
logger=None

class Result(enum.Enum):
    Unknown = None
    Less = -1
    Equal = 0
    Greater = 1

class DefaultComparator(traversal.Visitor):
    a: serializable.Serializable
    b: serializable.Serializable    
    result: Result
    
    def __init__(self, a: serializable.Serializable, b: serializable.Serializable):
        self.a = a
        self.b = b
        self.result = Result.Unknown

    @method(logger=logger)
    def begin(self, obj: serializable.ExpectedType, parent_prop_name: str = None) -> Result:
        self.result = Result.Equal
        return self.result

    @method(logger=logger)
    def end(self, obj: serializable.ExpectedType) -> Result:
        return self.result

    @method(logger=logger)
    def verbatim(self, 
        data_type: typing.Type, 
        target: serializable.Serializable, 
        get_value: typing.Callable [ [serializable.Serializable], typing.Any ],
        set_value: typing.Callable [ [serializable.Serializable, typing.Any ], None],
        get_prop_names: typing.Callable [ [], typing.Set[str] ]
    ) -> Result:
        if self.result != Result.Equal:
            return self.result
        a_prop = get_value(self.a)
        b_prop = get_value(self.b)
        if not a_prop is None and b_prop is None:
            self.result = Result.Less
        elif a_prop is None and not b_prop is None:
            self.result = Result.Greater
        elif a_prop is None and b_prop is None:
            pass
        elif a_prop < b_prop:
            self.result = Result.Less
        elif a_prop > b_prop:
            self.result = Result.Greater
        return self.result

    @method(logger=logger)
    def primitive(self, data_type: typing.Type, target: serializable.Serializable, prop_name: str, fromString: typing.Callable[ [str], PropType ] = None) -> Result:
        if self.result != Result.Equal:
            return self.result
        a_has_prop = hasattr(self.a, prop_name)
        b_has_prop = hasattr(self.b, prop_name)
        if not a_has_prop and not b_has_prop:
            return self.result
        if not a_has_prop and b_has_prop:
            self.result = Result.Less
        if a_has_prop and not b_has_prop:
            self.result = Result.Greater
        a_prop = getattr(self.a, prop_name)
        b_prop = getattr(self.b, prop_name)
        if not a_prop is None and b_prop is None:
            self.result = Result.Less
        elif a_prop is None and not b_prop is None:
            self.result = Result.Greater
        elif a_prop is None and b_prop is None:
            pass
        elif a_prop < b_prop:
            self.result = Result.Less
        elif a_prop > b_prop:
            self.result = Result.Greater
        return self.result

    @method(logger=logger)
    def compare_elements(self, a_prop: serializable.Serializable, b_prop: serializable.Serializable):
        if a_prop.get_global_id() is None and b_prop.get_global_id() is None:
            sub = DefaultComparator(a_prop, b_prop)
            a_prop.marshal(sub)
            if sub.result != Result.Equal:
                self.result = sub.result
        elif a_prop.get_global_id() < b_prop.get_global_id():
            self.result = Result.Less
        elif a_prop.get_global_id() > b_prop.get_global_id():
            self.result = Result.Greater

    @method(logger=logger)
    def scalar(self, element_builder: construction.Factory, target: serializable.Serializable, prop_name: str) -> Result:
        if self.result != Result.Equal:
            return self.result
        a_has_prop = hasattr(self.a, prop_name)
        b_has_prop = hasattr(self.b, prop_name)
        if not a_has_prop and not b_has_prop:
            return self.result
        if not a_has_prop and b_has_prop:
            self.result = Result.Less
        if a_has_prop and not b_has_prop:
            self.result = Result.Greater
        a_prop = getattr(self.a, prop_name)
        b_prop = getattr(self.b, prop_name)
        if not a_prop is None and b_prop is None:
            self.result = Result.Less
        elif a_prop is None and not b_prop is None:
            self.result = Result.Greater
        elif a_prop is None and b_prop is None:
            pass
        else:
            self.compare_elements(a_prop, b_prop)
        if self.result == Result.Unknown:
            self.result = Result.Equal
        return self.result

    @method(logger=logger)
    def array(self, element_builder: construction.Factory, target: serializable.Serializable, prop_name: str) -> Result:
        if self.result != Result.Equal:
            return self.result
        a_has_prop = hasattr(self.a, prop_name)
        b_has_prop = hasattr(self.b, prop_name)
        if not a_has_prop and not b_has_prop:
            return self.result
        if not a_has_prop and b_has_prop:
            self.result = Result.Less
        if a_has_prop and not b_has_prop:
            self.result = Result.Greater
        a_len = len(getattr(self.a, prop_name))
        b_len = len(getattr(self.b, prop_name))
        if a_len < b_len:
            self.result = Result.Less
        if a_len > b_len:
            self.result = Result.Greater
        for index in range(0, a_len):
            if self.result == Result.Unknown:
                self.compare_elements(getattr(self.a, prop_name)[index], getattr(self.b, prop_name)[index])
        if self.result == Result.Unknown:
            self.result = Result.Equal
        return self.result

    @method(logger=logger)
    def map(self, key_type: typing.Type, element_builder: construction.Factory, target: serializable.Serializable, prop_name: str) -> Result:
        if self.result != Result.Equal:
            return self.result
        a_has_prop = hasattr(self.a, prop_name)
        b_has_prop = hasattr(self.b, prop_name)
        if not a_has_prop and not b_has_prop:
            return self.result
        if not a_has_prop and b_has_prop:
            self.result = Result.Less
        if a_has_prop and not b_has_prop:
            self.result = Result.Greater
        a_prop = getattr(self.a, prop_name)
        b_prop = getattr(self.b, prop_name)
        if list(a_prop.keys()) < list(b_prop.keys()):
            self.result = Result.Less
        elif list(a_prop.keys()) > list(b_prop.keys()):
            self.result = Result.Greater
        a_keys = set([ key for key in a_prop.keys() ])
        b_keys = set([ key for key in b_prop.keys() ])
        if a_keys < b_keys:
            self.result = Result.Less
        elif a_keys > b_keys:
            self.result = Result.Greater
        for key in a_keys:
            if self.result == Result.Unknown:
                self.compare_elements(getattr(self.a, prop_name)[key], getattr(self.b, prop_name)[key])
        if self.result == Result.Unknown:
            self.result = Result.Equal
            
        return self.result

@function(logger=logger)
def cmp(a: serializable.Serializable, b: serializable.Serializable):
    if a.__class__.__qualname__ < b.__class__.__qualname__: return -1
    if a.__class__.__qualname__ > b.__class__.__qualname__: return 1

    if(hasattr(a, '__cmp__')):
        return getattr(a, '__cmp__')(b)

    if a is None and b is not None:
        return -1
    elif a is not None and b is None:
        return 1
    elif a is None and b is None:
        return 0

    comparator = DefaultComparator(a, b)
    a.marshal(comparator)
    if comparator.result == Result.Less:
        return -1
    elif comparator.result == Result.Equal:
        return 0
    elif comparator.result == Result.Greater:
        return 1
    raise RuntimeError('Objects cannot be compared')
